### ISSUE TITLE
Add ability to get chunks data before rendered to tags

### ISSUE THREAD

**Issue by NateWr** [mindplay-dk/php-vite#3]
I'm using this to integrate with another PHP application that has it's own method for declaring and loading `<link>` and `<script>` tags. Would you be open to adding a mechanism that returned the chunks, with the file type logic assigned?

I would suggest a new public function `createChunks()` that returned an assoc array:

```
[
  'modulepreload': [url, ...],
  'preload': [url, ...],
  'stylesheet': [url, ...],
  'script': [url, ...],
]
```

Happy to provide a PR but wanted to check with you first.

**Comment by mindplay-dk** [comment]
So you don't want the tags created for you?

I mean, that's mainly what the library does.

If you don't want that (and don't even want the type-safety) you can probably just `json_decode` the file yourself?

It sounds like you want a framework integration, in a framework where apparently you don't have the freedom to render your own tags in `<head>`? It sounds like an odd requirement. I'd prefer not to add any complexity unless there is a clear benefits.

Are you sure you're trying to solve the right problem? 😅

**Comment by NateWr** [comment]
Yeah, fair enough. I can write my own integration. I was just hoping to piggy-back on the file-type preload definitions you've set up and, I imagine, will eventually modify. :smile: 

(In my case, I'm creating a solution for themes for an [open source project](https://github.com/pkp/ojs). Injecting the tags into the template is just one more step I want to automate for theme creators.)

**Comment by mindplay-dk** [comment]
It sounds like you'd have an easier time for your particular use-case by just copying and modifying? 🙂

The file-type preload definitions are unlikely to change much - these were referenced from the MDN page on preloads and it only changes if the web standards change, which is fairly rare.

**Comment by mindplay-dk** [comment]
If you'd prefer to add this feature rather than maintaining your own project, please feel free to submit a PR, and I will consider it.

If not, I'll consider this issue as closed for now.

**Comment by NateWr** [comment]
I've gone ahead with my own package, but thanks for `php-vite`. It was a useful example to follow.

**Comment by mindplay-dk** [comment]
That's cool. 🙂

It looks like the main difference with your package is it doesn't render anything, and instead has an extra degree of abstraction in the form of an interface? That's actually a good approach. I might take that under consideration.

FYI, the code you copied is licensed under MS-RL:

https://github.com/mindplay-dk/php-vite/blob/d988b75e9cbf5592bf8e5933bb610f80031826a5/LICENSE#L28

Also note that you removed attribution when you removed the source history and `composer.json`. What I usually do when copying just a single file, is place an `@link` in a doc-block in the file, stating where it was copied from. Of course, the README is fine as well, attribution just needs to be somewhere. 🙂

(I know this license is somewhat unusual, but it's actually the same as GPL in those two regards. The main difference from e.g. MIT or GPL is you can't take it offline and make it private. A former employer took all of my open source projects offline and made them private, hence the somewhat unusual choice of license.)

**Comment by NateWr** [comment]
Sorry about that. I didn't notice the license restriction. I'll have to look into its compatibility with GPL, as the ability to use this package in proprietary code will be important for my use case. (These are not big companies but typically small publishers at university libraries or scholarly associations, often in the developing world.)

I totally understand your position, though. I believe `ManifestFile` is the only file I've copied from your library. Regardless of whether I need to rewrite from scratch to address licensing issues, I'd still plan to credit your work as inspiration.

>  instead has an extra degree of abstraction in the form of an interface?

Yes, this is the template library for the open source software I am integrating with (it's a Smarty wrapper).

**Comment by mindplay-dk** [comment]
> I'll have to look into its compatibility with GPL, as the ability to use this package in proprietary code will be important for my use case.

There are no restrictions about using the library in proprietary projects - that is the reason I chose this license. There are more commonly used licenses of this type, but they would require your project to be open source, which obviously isn't realistic, or would require you to disclose (e.g. in an about dialog) in your application that you're using the library, which I also did not want.

To the best of my understanding, the only difference with this license from e.g. MIT, is you can't take the library itself offline - you have to share it with the rest of us. I honestly thought this was self-explanatory and generally the spirit of open source, but my past employer clearly didn't think so. 😌

I honestly would have preferred the MIT license, which I've always used for most of my projects, because that's one that most people know. I decided to go with this one since it addresses a problem I had, and because it's still very short and easy to read. Similar licenses (with additional constraints like affecting your project license or enforcing disclosure) are extremely long and complex.

The other option was a modified MIT license, but I'm not a lawyer, so that might be worse. 😅

**Comment by NateWr** [comment]
That all makes sense to me. The stumbling block, I think, is that it is listed as [not compatible with GPL](https://www.gnu.org/licenses/license-list.html), which is what my theme code will likely be licensed under. I'm also not a lawyer and so it's difficult to evaluate the implications here. I'm really sympathetic to your reasoning, but the overhead from licensing ambiguity is a challenge.

**Comment by mindplay-dk** [comment]
In the context of libraries, being "incompatible" just means:

* You cannot take MS-RL licensed code and include it in a GPL-licensed library
* You cannot take MS-RL licensed code and relicense it under GPL
* You cannot create a derivative library that tries to be both MS-RL and GPL

I would suggest you take a deep breath and actually read the MS-RL license - this will take about 5-10 minutes for most people. You won't need a lawyer, and it's very easy to understand.

I know the GPL is much more common, but it's also more than 10x longer than MS-RL, arguably much harder to read - and lawyers, to this day, are still debating whether or not linking to a GPL licensed library creates a derivative work.

https://www.perplexity.ai/search/does-linking-to-a-gpl-library-gFYyD09xRmmbnmTVr6imTA#0

The bottom line is that, in my experience, GPL is too confusing and/or restrictive, MIT is too permissive - and if you're only willing to consider other "popular" licenses, those are all generally either too restrictive or too complex.

I honestly wish the MS-RL license didn't carry the MS brand. People automatically think it's "bad" because it was created by Microsoft. But, after the MIT license, it's the simplest and least confusing open source license I've been able to find.

Needing to understand or explain the choice is obviously a huge drawback for both of us.

But my only real option is changing the license to MIT, which, while great for some things, isn't a reciprocal license.

I'm guessing that's also the main reason you chose GPL and not MIT?

**Comment by mindplay-dk** [comment]
@NateWr I've decided to change the license to MPL-2.0.

It's incredibly long and complex, and I'm actually still somewhat uncomfortable using a license that I can't personally read and understand first-hand.

However, having consulted a number of sources, the MPL-2.0 seems like a 10x longer, unreadable MS-RL, that people *trust*. 😅

I don't think it's great that we use licenses we can only understand *indirectly*, and therefore only trust *implicitly* - but in this case, with the apparent confusion around MS-RL (and potentially distrust of the MS brand) it seems like the least bad choice.

You are of course free to use the code as it was already published under MS-RL, but I will change the license to MPL-2.0 now, so you can use it under that license, if you prefer. 🙂

**Comment by NateWr** [comment]
@mindplay-dk I had to pivot to another project but I'll be back on this soon. Sorry the licensing is such a pain. Since I'm not an expert, I have to rely more on convention than anything legal, which is a shame! MPL-2.0 seems to address the problem for me, since the code will be distributed as part of a larger GPL work.Thanks!  :+1:

**Comment by mindplay-dk** [comment]
@NateWr no problem - my licensing issues started long before you brought this up! I've been trying to figure this out for *years*. Trying out MS-RL was a failed experiment - now that I understand MPL-2.0 a little better, I will most likely use that for other projects going forward. 🙂👍

Note that GPL is compatible with MPL-2.0, which means you *can* distribute my code under GPL. 👍

The only thing missing in your project is attribution. 🙂

**Comment by NateWr** [comment]
After a long delay, I finally got around to writing a bit of docs and crediting you appropriately. Are you happy with this in the readme?

https://github.com/NateWr/pkp-vite?tab=readme-ov-file#credit

**Comment by mindplay-dk** [comment]
@NateWr totally fine, thanks! 🙂
